processing/src/main/java/org/apache/druid/math/expr/Function.java

```diff
@@ -3779,15 +3781,64 @@ public ExpressionType getOutputType(Expr.InputBindingInspector inspector, List<E
     }
 
     @Override
-    ExprEval doApply(ExprEval lhsExpr, ExprEval rhsExpr)
+    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
     {
+      final ExprEval lhsExpr = args.get(0).eval(bindings);
+      final ExprEval rhsExpr = args.get(1).eval(bindings);
+
       final Object[] array1 = lhsExpr.asArray();
-      final Object[] array2 = rhsExpr.asArray();
-      return ExprEval.ofLongBoolean(Arrays.asList(array1).containsAll(Arrays.asList(array2)));
+      if (array1 == null) {
+        return ExprEval.ofLong(null);
+      }
+      ExpressionType array1Type = lhsExpr.asArrayType();
+
+      if (rhsExpr.isArray()) {
+        final Object[] array2 = rhsExpr.asArray();
+
+        if (array2 == null) {
+          return ExprEval.ofLongBoolean(false);
+        }
+        final Set<Object> set = array1Type.isPrimitiveArray()
```

Comment:
ArrayContains should be an ExprMacro, since the rhs expr is often literal, and we don't want to be rebuilding a set for each row (building sets isn't super-cheap).